### PR TITLE
bpo-29546: Set path on ImportError for from ... import

### PR DESCRIPTION
**Pull request by Carreau**
And `path` as well when existing.

See bpo-29546

This is a step toward providing better error messages in case of
from-import. Barry Warsaw Proposed:

   cannot import name {name} from {module} ({path})

But that's probably going to trigger more discussions that filling in
already existing fields.

--- 
I'm not familiar with C-internal of CPython, So let me know if I'm doing things wrong.

### CODE REVIEW

File: Python/ceval.c
Comment (brettcannon) [inline Python/ceval.c]: Looks like there's a missing space between the final `)` and `{`.

**Comment by brettcannon** [comment]
I have not dug into the surrounding code, but the general approach LGTM.

File: Python/ceval.c
Comment (Carreau) [reply Python/ceval.c]: Dang. Fixed.

**Comment by brettcannon** [comment]
I also just realized there's no test.

**Review by brettcannon — CHANGES_REQUESTED** [review]
Please add a test.

**Comment by Carreau** [comment]
> I also just realized there's no test.

Yes I realized that as well :-) Will do .

File: Lib/test/test_import/__init__.py
Comment (brettcannon) [inline Lib/test/test_import/__init__.py]: We don't include docstrings on tests (and we PEP 8 doesn't allow starting a docstring with a newline).

File: Lib/test/test_import/__init__.py
Comment (brettcannon) [inline Lib/test/test_import/__init__.py]: Could you do an `assertEqual` check using `os.__file__`?

File: Lib/test/test_import/__init__.py
Comment (brettcannon) [inline Lib/test/test_import/__init__.py]: The custom message isn't necessary as what the test is checking is obvious (same goes for all of the other messages).

File: Lib/test/test_import/__init__.py
Comment (brettcannon) [inline Lib/test/test_import/__init__.py]: Any reason you can't use `with self.assertRaises()`?

File: Lib/test/test_import/__init__.py
Comment (brettcannon) [inline Lib/test/test_import/__init__.py]: Unnecessary blank line.

**Review by brettcannon — CHANGES_REQUESTED** [review]

File: Lib/test/test_import/__init__.py
Comment (Carreau) [reply Lib/test/test_import/__init__.py]: I just forgot you could access the exception after the fact with `contextmanager.exception` so I indeed I can ans should use it.

File: Lib/test/test_import/__init__.py
Comment (Carreau) [reply Lib/test/test_import/__init__.py]: Sure.

File: Lib/test/test_import/__init__.py
Comment (brettcannon) [inline Lib/test/test_import/__init__.py]: Typo" cannonical" -> "canonical"

File: Lib/test/test_import/__init__.py
Comment (brettcannon) [inline Lib/test/test_import/__init__.py]: Should there be a test for `path`, even if it's just checking it isn't `None`?

File: Lib/test/test_import/__init__.py
Comment (brettcannon) [inline Lib/test/test_import/__init__.py]: Could you add a comment like `# _warning has no path as it's a built-in module.`? I keep forgetting why this specific case doesn't have a path. 😄

**Review by brettcannon — CHANGES_REQUESTED** [review]

File: Lib/test/test_import/__init__.py
Comment (Carreau) [reply Lib/test/test_import/__init__.py]: It could, but that's taken care of by functions above. Here I'm just checking that it's not `os.path`. 
Anyway while I'm at it I will add an assertIsNotNone.

File: Lib/test/test_import/__init__.py
Comment (Carreau) [reply Lib/test/test_import/__init__.py]: Too early in the bootstrap process ? 
Anyway. Done.

File: Lib/test/test_import/__init__.py
Comment (Carreau) [reply Lib/test/test_import/__init__.py]: I need to find a linter that understand both programming and english.

**Comment by berkerpeksag** [comment]
BTW, please rebase your branch so we can get rid of these Codecov comments :)

**Comment by Carreau** [comment]
> BTW, please rebase your branch so we can get rid of these Codecov comments :)

Oh gosh 71 was merged ? Thanks you! and done.

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
LGTM! Thanks for sticking through it, @Carreau !

Please add yourself to `Misc/ACKS` if you're not already there and add an entry into `Misc/NEWS` like:
```
bpo-29546: Set the 'path' attribute on ImportError for ``from ... import ...``.
```

**Comment by Carreau** [comment]
> changed the title from bpo-29546: Set name on ImportError for from ... import to bpo-29546: Set path on ImportError for from ... import

Well actually good point that's setting both name and path. Name was not set either if that matters. 



> Please add yourself to Misc/ACKS if you're not already there and add an entry into Misc/NEWS like:

Sure !

**Comment by brettcannon** [comment]
Thanks! As soon as Travis is green I'll merge.

**Comment by Carreau** [comment]
> Thanks! As soon as Travis is green I'll merge.

Thanks. I don't know how disturbing the migration has been from your end (and other core dev). 
From my side being able to use git and PRs was great. 

Many kudos for having worked on the migration for so long.

**Comment by brettcannon** [comment]
Merged! Thanks again!

And I'm glad that the new workflow worked out for you! As for me, the new workflow let me review this PR throughout the day when I had time and do it entirely in the browser so I didn't have to remember to commit it when I got home (if I had to cherry-pick that would not have been the case, but I could at least merge now and cherry-pick later if need be).

**Comment by Carreau** [comment]
> if I had to cherry-pick that would not have been the case

Hopping you will get the cherry-pick bot soon. Thanks for the time you've put into this. 
I probably owe you a "orange juice, root beer, or Canada Dry," when we meet IRL.

**Comment by brettcannon** [comment]
@Carreau the bot would be great, but at worst me or someone else can write a Python script probably to automate it locally if necessary (which reminds me, I should write that idea down over on core-workflow).

**Comment by Carreau** [comment]
> which reminds me, I should write that idea down over on core-workflow

You did https://github.com/python/core-workflow/issues/8 (Bot to automatically generate cherry-picking PRs)

**Comment by warsaw** [comment]
Wow, I'm very pleasantly impressed at how quickly my bug/idea went to (nearly) implemented through the new workflow.  Kudos!

**Comment by Carreau** [comment]
> Wow, I'm very pleasantly impressed at how quickly my bug/idea went to (nearly) implemented through the new workflow. Kudos!

Happy to repay a bit for all the work you did. 

It's also much easier to decide to implement if a core dev have pitched in the idea, as it has more chance to be accepted. Thanks both !

**Pull request merged**
